feat: Add client-side logging support

ajborley opened the pull request:
- Adds the websocket listener to the server.
- Adds support on the client to send logging events through the
websocket via a new `useLogger` hook.

Contributes to: #42

Signed-off-by: Andrew Borley <borley@uk.ibm.com>

ajborley committed:
feat: Add client-side logging support

- Adds the websocket listener to the server.
- Adds support on the client to send logging events through the
websocket via a new `useLogger` hook.

Contributes to: #42

Signed-off-by: Andrew Borley <borley@uk.ibm.com>

ajborley commented:
Draft PR to check if folks think I'm on the right track... still working on how to test this all.

ajborley committed:
fix: Add server-side logging WebSocket tests

Contributes to: #42

Signed-off-by: Andrew Borley <borley@uk.ibm.com>

pmuir commented on client/Hooks/useLogger/Hook.ts:
```diff
@@ -0,0 +1,94 @@
+export const useLogger: (string) => (LogLevel, string) => void = (
+  componentName: string
+) => {
+  // TODO: store the logger websocket in context, so you get one for the whole UI, rather than one every time you call useLogger.
```
Feels like we need this before we can merge?

ajborley replied:
Agreed, and the same for the other TODOs (https://github.com/strimzi/strimzi-ui/pull/85/files#diff-3e52b502269a41a2c06ce466fd7d975b726b3557bc98a54d84b74b2bf958f1baR9-R12)

nictownsend replied:
Is this still blocking merge? Feels like yes?

pmuir replied:
I think it does. Opening a websocket for every log request defeats the purpose of using a websocket.

ajborley replied:
Latest commit resolves this via context

nictownsend commented on client/Hooks/useLogger/Hook.ts:
```diff
@@ -0,0 +1,94 @@
+    }
+  };
+
+  return logger;
```
Could we have the standard `log` `debug` etc as curried methods instead of passing the level in whenever you call the logger method?

```
const genLog = (logLevel) => (message) => {
  ...
 buffer.push({logLevel, message});
  ...
};


return {
  info : genLog("info);
  ...
}
```

ajborley replied:
Good idea - done.

nictownsend commented on client/Hooks/useLogger/Hook.ts:
```diff
@@ -0,0 +1,94 @@
+      // Add a new new message to the buffer
+      loggerMessageBuffer.push({
+        clientTime: Date.now(), // client timestamp in ms
+        clientID: defaultClientID, // client ID
```
UUID? Or something sent up from server when the socket opens?

ajborley replied:
Yep - have picked up the `uuid` package.

nictownsend commented on server/log/router.ts:
```diff
@@ -14,8 +21,51 @@ export const LogModule: UIServerModule = {
+  data
+) => {
+  if (typeof data === 'string') {
+    try {
```
https://github.com/DefinitelyTyped/DefinitelyTyped/blob/c1c23b2dcaf1a419992379a2b5bf7e14fe829f7a/types/ws/index.d.ts#L118 - any chance we could end up with buffers? Is it client that decides the format?

ajborley replied:
Yep - the `useLogger` hook will always send strings so this should be fine.

nictownsend commented on server/log/router.ts:
```diff
@@ -14,8 +21,51 @@ export const LogModule: UIServerModule = {
+            clientLogEvent.msg
+          );
+        } else {
+          logger.debug(clientLogEvent, clientLogEvent.msg);
```
https://github.com/pinojs/pino/blob/HEAD/docs/api.md#logging-method-parameters - looks like we don't need second param as our event contains `msg` field

ajborley replied:
Good spot, have removed.

nictownsend commented on server/log/router.ts:
```diff
@@ -14,8 +21,51 @@ export const LogModule: UIServerModule = {
+const handleClose: (
+  logger: Logger
+) => (code: number, reason: string) => void = (logger) => (code, reason) =>
+  logger.debug(`WebSocket listener closed. (${code}) ${reason}`);
```
Can we log out a request ID / client ID?

ajborley replied:
The close event only receives the code and reason string: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/ws/index.d.ts#L86 so we would need to encode the client ID in the reason string. I will look into this when I hook into the context, as we may want to explicitly close the logging websocket on client unload.

nictownsend commented on server/log/router.ts:
```diff
@@ -14,8 +21,51 @@ export const LogModule: UIServerModule = {
   },
 };
+
+const handleMessage: (logger: Logger) => (data: Data) => void = (logger) => (
```
Pedantic, but can you rename to `messageHandler` (as it's not itself handling it, but the return function.

ajborley replied:
Done

nictownsend commented on server/log/router.ts:
```diff
@@ -14,8 +21,51 @@ export const LogModule: UIServerModule = {
+        ws.on('message', handleMessage(logger));
+        ws.on('close', handleClose(logger));
+      } else {
+        // Return 426 Upgrade Required if this isn't a websocket request
```
Is this just belts and braces on the error case? Could return `404` as it's not a valid call for non web?

ajborley replied:
We expect HTTP GET calls to the `/log` endpoint to be websocket upgrade requests. If we get a request that is not an upgrade, then we return the [426](https://httpstatuses.com/426) response, as we don't serve any other http traffic from `/log`.

nictownsend commented on server/main.ts:
```diff
@@ -69,6 +70,18 @@ loadConfig((loadedInitialConfig) => {
+      req.isWs = true;
+      res.ws = ws;
+      // call the express app as usual
+      expressApp(req, res);
```
How does `expressApp` know which endpoint to call? Is it simply that we're listening on `host:port/*` so any websocket request will be upgraded, but then sending messages will fail unless it's on a path that has `isWs` support?

ajborley replied:
Yep, that's right. The code in main.ts (https://github.com/ajborley/strimzi-ui/blob/42-client-side-logging/server/main.ts#L74-L90) watches for websocket upgrade requests, and adds the `isWs` boolean and `ws` websocket to the req and res respectively. So currently, any endpoint that receives an upgrade request will set up a websocket, but only modules that implement the `websocket.onmessage`, etc handlers will actually do anything.

I expect this is not what we want - we should probably add something to the module configuration that can be used to decide whether the websocket upgrade request is permitted or not. What do you think?

nictownsend replied:
I wonder if we need a break out session on this.

It seems like we're possibly shoehorning in the `log` module into express, when actually this is a different paradigm. My concern is that by appending `ws` to `req` we lose reference to it outside of the initial request.

Without extending express or using old express middleware - it feels like we need the ability to control "websocket" modules too - maybe, by having the concept of attaching multiple endpoints (https://www.npmjs.com/package/ws#multiple-servers-sharing-a-single-https-server) - we then destroy socket if it's not on a path that's acceptable?

We can then have "express" modules and "ws" modules (with a similar `addModule` for ws that takes in a logger)

nictownsend commented on server/test_common/commonServerSteps.ts:
```diff
@@ -83,4 +119,83 @@ When(
+  stepWithWorld(async (world) => {
+    const { websocket } = world;
+    if (websocket.onclose != null) {
+      expect((<jest.Mock>websocket.onclose).mock.calls.length).toBe(1);
```
`toHaveBeenCalled` and related - https://jestjs.io/docs/en/expect#tohavebeencalled - is a neater API than `mock.calls.x`

ajborley replied:
Forgot all about the `toHaveBeenCalled*` API.  :) Updated

ajborley committed:
fix: Add client-side useLogger hook tests

Contributes to: #42

Signed-off-by: Andrew Borley <borley@uk.ibm.com>

ajborley committed:
fix: Address review comments

- Address initial review comments from @nictownsend

Contributes to: #42

Signed-off-by: Andrew Borley <borley@uk.ibm.com>

matthew-chirgwin commented:
Requesting (re) reviews please @nictownsend / @pmuir !

nictownsend commented on client/Hooks/useLogger/README.md:
```diff
@@ -0,0 +1,45 @@
+
+## API
+
+### `userLogger(componentName: string)`
```
```suggestion
### `useLogger(componentName: string)`
```

ajborley replied:
Fixed

nictownsend commented on client/Hooks/useLogger/useLogger.steps.tsx:
```diff
@@ -0,0 +1,131 @@
+
+After(() => {
+  sentMessagesCount = 0;
+  mockWebSocketServer.close();
```
May be undefined if the `Given` was never called

ajborley replied:
Check added

nictownsend commented on client/Hooks/useLogger/useLogger.steps.tsx:
```diff
@@ -0,0 +1,131 @@
+  verifiedClient = true;
+});
+
+Given('a logging WebSocket server', () => {
```
Worry is that if jest runs tests in parallel, or someone else uses this `Given` we'll get port clashes. But if this is fixed via the context change as a later commit (or new PR) I'm less worried. Wonder if when you add in the `context` we can supply a mock ws client via context instead of having to start up real servers in the test.

ajborley replied:
The `new Server(..)` line here is a [`mock-socket`](https://www.npmjs.com/package/mock-socket) server, which just hooks mock-clients to mock-servers using the supplied server URL as a key. No real servers are being started here :)

nictownsend commented on client/Hooks/useLogger/useLogger.steps.tsx:
```diff
@@ -0,0 +1,131 @@
+  }
+};
+Then(loggingMessagesAreReceived, loggingMessagesAreReceivedFn);
+And(loggingMessagesAreReceived, loggingMessagesAreReceivedFn);
```
Eugh, that's unfortunate syntax :(

ajborley replied:
Agreed, ugly. Couldn't spot any alternative way to use `And` for the same code as a `Given`, `When` or `Then`. If you have any good ideas... :)

nictownsend commented on server/main.ts:
```diff
@@ -69,6 +70,26 @@ loadConfig((loadedInitialConfig) => {
+        req.isWs = true;
+        res.ws = ws;
+        // call the express app as usual
+        expressApp(req, res);
```
Please see previous comment re breakout - not sure how safe this is for GC purposes that we lose reference to the ws (to be able to close it etc)

matthew-chirgwin replied:
@ajborley / @nictownsend has this chat happened yet? Would be keen to participate. If the `ws` was placed in `res.locals`, it will be scoped to the life of the request (ie socket lifetime). Express will then discard the `locals`, `ws` included. Thoughts?

ajborley replied:
I think this will be superseded when Matt's websocket handling changes go in, so I've left it as-is for now.

matthew-chirgwin replied:
Yep - I will update the approach in my next PR

dlabaj commented on client/Hooks/useLogger/Hook.ts:
```diff
@@ -0,0 +1,125 @@
+  | 'debug'
+  | 'trace';
+
+interface loggerMessage {
```
Are these all required?

ajborley replied:
These are all the levels available in the `pino` library used on the server. Only those at the `pino` logging level will actually get emitted to the log.

dlabaj commented on client/Hooks/useLogger/Hook.ts:
```diff
@@ -0,0 +1,125 @@
+  useEffect(() => {
+    // Check if the component should be logged
+    if (defaultLoggingRegex.test(componentName)) {
+      const loggerWebSocket = loggerWebSocketRef.current;
```
Did we want to declare these types instead of using inference?

ajborley replied:
This code has been updated, hopefully the types are clearer now.

dlabaj reviewed (approved):
A few comments otherwise looks good.

pmuir reviewed (changes requested):
TODOs need resolving, otherwise LGTM

matthew-chirgwin reviewed (commented):
LGTM - re the TODOs, the first stage of config work has now landed - would it be possible to get the various TODOs re config usage to be updated to use that, which I can then build upon in #99 ?

ajborley committed:
feat: Use the LOGGING query param to select which components are logged

Contributes to: #42

Signed-off-by: Andrew Borley <borley@uk.ibm.com>

ajborley committed:
feat: Store websocket and messagebuffer in context

Contributes to: #42

Signed-off-by: Andrew Borley <borley@uk.ibm.com>

JQrdan commented on client/Contexts/Logging/Logging.types.ts:
```diff
@@ -0,0 +1,24 @@
+  | 'warn'
+  | 'info'
+  | 'debug'
+  | 'trace';
```
Would this be nicer as an enum?

ajborley replied:
Yep, have updated.

JQrdan commented on client/Contexts/Logging/README.md:
```diff
@@ -0,0 +1,3 @@
+# Logging
+
+This context will store the logging WebSocket cient connection and the buffer of logging messages used in the (useLogger)[../../Hooks/useLogger] hook. This allows the useLogger hook to use a single WebSocket connection for all logging calls.
```
```suggestion
This context will store the logging WebSocket client connection and the buffer of logging messages used in the (useLogger)[../../Hooks/useLogger] hook. This allows the useLogger hook to use a single WebSocket connection for all logging calls.
```

ajborley replied:
Done

JQrdan commented on client/Hooks/useLogger/Hook.ts:
```diff
@@ -0,0 +1,132 @@
+ */
+const shouldLogComponent: (componentName: string) => boolean = (
+  componentName: string
+) => {
```
You can streamline typing like this:

```suggestion
const shouldLogComponent = (
  componentName: string
): boolean => {
```

then `shouldLogComponent` will infer the type from the right side. It saves having to write parameters and their types twice, which becomes painful when you change them

ajborley replied:
Nice! Thanks for that, makes it much tidier :)

JQrdan commented on client/Hooks/useLogger/Hook.ts:
```diff
@@ -0,0 +1,132 @@
+  return shouldLogComponent;
+};
+
+const getWebSocketAddress: () => string = () => {
```
```suggestion
const getWebSocketAddress = (): string => {
```

ajborley replied:
Done

JQrdan commented on client/Hooks/useLogger/Hook.ts:
```diff
@@ -0,0 +1,132 @@
+
+export const useLogger: (componentName: string) => LoggerType = (
+  componentName: string
+) => {
```
```suggestion
export const useLogger = (
  componentName: string
): LoggerType => {
```

ajborley replied:
Done

JQrdan commented on client/Hooks/useLogger/useLogger.steps.tsx:
```diff
@@ -0,0 +1,167 @@
+});
+
+Given(/^a(?: (secure))* logging WebSocket server$/, (isSecure) => {
+  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+  // @ts-ignore
```
Not a review comment but just wanted to say this is amusing 😂

ajborley replied:
I know :)

JQrdan commented on client/Hooks/useLogger/useLogger.steps.tsx:
```diff
@@ -0,0 +1,167 @@
+  const wrapper = ({ children }) => (
+    <LoggingProvider>{children}</LoggingProvider>
+  );
+  // eslint-disable-next-line react-hooks/rules-of-hooks
+  const renderedHook = renderHook(() => useLogger('TestComponent'), {
+    wrapper,
+  });
```
I'm curious why you need this eslint comment? I though renderHook behaved so that it should abide by all the rules

ajborley replied:
Without it the `renderHook` line fails the linter, even though I'm following the [renderHook examples](https://react-hooks-testing-library.com/usage/basic-hooks) (I'm guessing the lint rules just aren't smart enough yet):
```
/strimzi-ui/client/Hooks/useLogger/useLogger.steps.tsx
  84:41  error  React Hook "useLogger" cannot be called inside a callback. React Hooks must be called in a React function component or a custom React Hook function  (react-hooks/rules-of-hooks)
```

JQrdan replied:
Very strange, in my code I have `const { result } = renderHook(() => useRouteConfig(input));` and it's happy

ajborley committed:
fix: Address review comments

 - Address @JQrdan's review comments

Contributes to: #42

Signed-off-by: Andrew Borley <borley@uk.ibm.com>

JQrdan reviewed (approved):
LGTM

ajborley committed:
fix: Avoid pino logger destructuring bug

 - Temporary fix to avoid the issue where destructuring the pino Logger
causes the destructured functions to fail due to `this` becoming
undefined.

Contributes to: #42

Signed-off-by: Andrew Borley <borley@uk.ibm.com>
